src/wallet/rpcwallet.cpp

```diff
@@ -1141,13 +1167,35 @@ UniValue addmultisigaddress(const JSONRPCRequest& request)
     if (!request.params[2].isNull())
         strAccount = AccountFromValue(request.params[2]);
 
+    int required = request.params[0].get_int();
+
+    // Get the public keys
+    const UniValue& keys = request.params[1].get_array();
+    std::vector<CPubKey> pubkeys;
+    pubkeys.resize(keys.size());
+    for (unsigned int i = 0; i < keys.size(); ++i) {
+        _createmultisig_addr_to_pubkey(pwallet, pubkeys.at(i), keys[i].get_str());
+        _createmultisig_hex_to_pubkey(pubkeys.at(i), keys[i].get_str());
+        if (!pubkeys.at(i).IsFullyValid()) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid public key: " + keys[i].get_str());
+        }
+    }
+
     // Construct using pay-to-script-hash:
-    CScript inner = _createmultisig_redeemScript(pwallet, request.params);
+    CScript inner = _createmultisig_redeemScript(required, pubkeys);
     CScriptID innerID(inner);
     pwallet->AddCScript(inner);
 
     pwallet->SetAddressBook(innerID, strAccount, "send");
-    return EncodeDestination(innerID);
+
+    UniValue result(UniValue::VOBJ);
+    if (IsDeprecatedRPCEnabled("addmultisigaddress")) {
```

Comment:
This can be made simpler:

```
    if (IsDeprecatedRPCEnabled("addmultisigaddress")) {
        // Old-style interface: just return the address
        return EncodeDestination(innerID);
    }

    // return an object containing the address and redeem script
    UniValue result(UniValue::VOBJ);
    result.pushKV("address", EncodeDestination(innerID));
    result.pushKV("redeemScript", HexStr(inner.begin(), inner.end()));
    return result;
```

Reply:
Done